Improve lmms::fastRand() and use it instead of std::rand()

rubiefawn opened the pull request:
- Combine overloads of `lmms::fastRand()` into one template
  - Remove `RAND_MAX` constant since it is not used outside `lmms::fastRand()` (and all places that did previously use it should have been calling `lmms::fastRand()` anyway)
  - This is also done to `lmms::fastLog10f()` for consistency, though it is not related to the goal of the PR
- Add `lmms::oneIn()` convenience function to replace occurences of `(fastRand() % chance == 0)`
- Replace uses of `std::rand()` with `lmms::fastRand()` for consistency
- Replace uses of `scale * lmms::fastRand() / RAND_MAX - offset` with `lmms::fastRand(min, max)`

```c++
// Old
int x = static_cast<int>(10.0f * rand() / RAND_MAX - 3.0f);

// New
int x = fastRand(-3, 7);
```

# Testing

- [x] Triple Oscillator noise sounds the same
- [x] Organic randomize preset button works
- [x] Multiple Peak Controllers can coexist
- [x] Sfxr randomize preset button works
- [x] SID plugin sounds the same
- [x] Vibed plugin sounds the same
- [x] `.ds` samples with noise sound the same
- [x] Undo/redo doesn't instantly explode
- [x] Mixer channel color randomization works
- [x] Clip color randomization works
- [x] Track color randomization works

rubiefawn committed:
Revert mistaken deprecation warning in LADSPA code

rubiefawn committed:
Change a sin() to std::sin() in Sfxr

rubiefawn committed:
Maybe fix silly template issues

rubiefawn committed:
Fix missing semicolon

templates are still busted tho

rubiefawn committed:
Remove type constraints

I'll revisit those later. C++ is weird

rubiefawn committed:
Actually commit warning fixes that I didn't stage

rubiefawn committed:
Time to stop screwing around with templates lol

Templates really would not give that much of a benefit here. I think I
was just having fun, but it's not working correctly with clang, so time
to get real.

rubiefawn committed:
Remove unneeded include

rubiefawn committed:
Replace most uses of std::rand() with fastRand()

rubiefawn committed:
Restore previously deleted FAST_RAND_MAX constant

Now that it's used in more than one place, it deserves to exist

rubiefawn committed:
Fix typos, warnings

rubiefawn committed:
Combine overloads of fastRand()

Combines the overloads for floating point types and integral types by
making the FAST_RAND_RATIO type conditional. It is now the same type as
the input range if the range is a floating point type, otherwise it
defaults to float. This removes the need for the FAST_RAND_MAX constant.

rubiefawn committed:
Include lmms_math.h for fastRand()

It appears I mistakenly deleted it during master merge

rubiefawn committed:
Move conditional type into concept

Also applies the "use T if it's a floating point type, else use float"
in order to combine overloads of fastLog10f() in a similar fashion

rubiefawn committed:
...well, it worked on godbolt lol

szeli1 commented on plugins/Organic/Organic.cpp:
```diff
@@ -346,27 +338,18 @@ float inline OrganicInstrument::waveshape(float in, float amount)
-
-		m_osc[i]->m_oscModel.setValue( intRand( 0, 5 ) );
+		m_osc[i]->m_volModel.setValue(fastRand(100));
+		m_osc[i]->m_detuneModel.setValue(fastRand(-5, +5));
```
```suggestion
		m_osc[i]->m_detuneModel.setValue(fastRand(-5, 5));
```

rubiefawn replied:
I included explicit positive signs at several callsites of `fastRand(-min, +max)` to communicate intent. I am willing to remove these, but only if others also disagree with my choice here.

szeli1 commented on include/lmms_math.h:
```diff
@@ -72,30 +75,36 @@ inline auto fraction(std::floating_point auto x) noexcept
-}
+
+//! @brief Returns a psuedorandom number between @p from and @p to, inclusive.
+inline auto fastRand(auto from, auto to) noexcept {	return from + fastRand(to - from); }
```
```suggestion
auto fastRand(auto from, auto to) noexcept {	return from + fastRand(to - from); }
```

szeli1 replied:
```suggestion
auto fastRand(auto from, auto to) noexcept { return from + fastRand(to - from); }
```
Remove tab also.

> Would you clarify your reasoning for suggesting this?

In the past, I was told to remove `inline`, since it is implicit.

rubiefawn committed:
Remove evil sneaky tab character

rubiefawn committed:
Forgot we do pointers like this now

inb4

messmerd reviewed (commented):
I haven't reviewed everything yet, but here's what I have so far

rubiefawn commented on include/lmms_math.h:
```diff
@@ -72,30 +75,36 @@ inline auto fraction(std::floating_point auto x) noexcept
-	static unsigned long next = 1;
-	next = next * 1103515245 + 12345;
-	return next / 65536 % 32768;
+	thread_local unsigned long s_next = 1;
```
I was under the impression that `thread_local` was similar to `static`, except `thread_local` allocates separate storage for each thread rather than only a single location for all threads to fight over. I made this change so that different threads each have their own `next`. Is this not the case?

rubiefawn commented on include/lmms_math.h:
```diff
@@ -72,30 +75,36 @@ inline auto fraction(std::floating_point auto x) noexcept
-}
+
+//! @brief Returns a psuedorandom number between @p from and @p to, inclusive.
+inline auto fastRand(auto from, auto to) noexcept { return from + fastRand(to - from); }
```
I will have to check since it's been awhile since I made this PR, but I remember two things:

First, there exist several places that call this overload with an `int`, so the omission of `std::floating_point` is intentional.

Second, there exists some places that call this overload with two different types, which would otherwise require an explicit cast if both arguments are required to be the same type. I'm less confident about this point so I'll have to double-check—I may be misremembering.

rubiefawn committed:
Address some code review suggestions

bratpeki commented on include/lmms_math.h:
```diff
@@ -77,25 +79,32 @@ inline auto absFraction(std::floating_point auto x) noexcept
 
-inline auto fastRand() noexcept
+//! @brief Returns a psuedorandom integer between 0 and 32767, inclusive.
+inline std::integral auto fastRand() noexcept
```
Apparently, there's `std::linear_congruential_engine`, which we may want to use for more clarity? I think the current code relies on overflow to do the LCG modulo operation.

bratpeki replied:
From [here](https://en.wikipedia.org/wiki/Linear_congruential_generator#Parameters_in_common_use):

```
modulus = 2^31
multiplier = 1103515245 <--- our number
increment = 12345 <--- our number
```

However, were using an `unsigned long` which is 64 bits, right? So my overflow comment shouldn't hold, because the modulus is different.

The worst part is that the code works when tested and this seems to have not changed the logic as it is in nightly.

This breaks my brain!

bratpeki replied:
Approving still, because it works and was probably there for _ages_.

rubiefawn commented on include/lmms_math.h:
```diff
@@ -38,6 +38,9 @@
 namespace lmms
 {
+using namespace std::numbers;
+using std::integral, std::floating_point, std::is_arithmetic_v, std::is_floating_point_v, std::conditional_t;
+
```
I've removed this line as requested, not sure why GitHub isn't marking this as outdated lol.

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-
-	s->m_startFreqModel.setValue( 0.5f+frnd(0.5f) );
-	s->m_minFreqModel.setValue(	s->m_startFreqModel.value()-0.2f-frnd(0.6f)	);
+	s->m_waveFormModel.setValue(fastRand(2));
```
`rnd(2)` returns an integer
`fastRand(2)` returns a float quantized by 1/32767 going upto 2
Right?

bratpeki replied:
Okay, took a more detailed look:

```cpp
inline std::integral auto fastRand() noexcept
{
	thread_local unsigned long s_next = 1;
	s_next = s_next * 1103515245 + 12345;
	return s_next / 65536 % 32768;
}
```

This returns in a range of [0, 32767].

```cpp
template<typename T> requires std::is_arithmetic_v<T>
inline auto fastRand(T range) noexcept
{
	constexpr auto FAST_RAND_RATIO = ...(1.0 / 32767);
	return fastRand() * range * FAST_RAND_RATIO;
}

```

- So the minimum of `fastRand(2)` is 0 * 2 * 1 / 32767 = 0.
- The maximum is 32767 * 2 * 1 / 32767 = 2.
- One before the maximum is 32766 * 2 * 1 / 32767 = 1.999939.

If we count on implicit conversion of floats to ints, then we just chop off the floating point bit, right?

That means:
- the chance for [0, 1.999939] (the chance we get 0 and 1) is 32766/32767 and
- the chance we get 2 is 1/32767.

The OG code has this distribution:
```
0 | 33.3%
1 | 33.3%
2 | 33.3%
```

Now, it's (in rough numbers):
```
0 | (32766/2) / 32767 = 49.99%
1 | (32766/2) / 32767 = 49.99%
2 | 1 / 32767 = 0.003%
```

I believe swapping the `fastRand(2)` for `fastRand() % 3` should solve this, since you get integers and the same 33.3% distribution of 0, 1 and 2.

rubiefawn replied:
The assumption that "If we count on implicit conversion of floats to ints, then we just chop off the floating point bit" doesn't hold. The rounding is not implicit (no conversion to `int` ever actually happens) and the rounding occurs in `AutomatableModel::fittedValue()` through `nearbyintf()`, which should be rounding to the nearest integer.

However, this still does point out a valid problem, which is that the distribution of integers isn't uniform. What it actually looks like is:

```
[0.0, 0.5) => 0 | 25%
[0.5, 1.5) => 1 | 50%
[1.5, 2.0] => 2 | 25%
```

Using an integer cast and modulo isn't the solution for a uniform distribution either (the OG code's distribution is not a clean 33.3%!) `[0, 32767] % x` is only uniform when `x` divides `32767`, which is not always true. Most of the time a few of the early numbers will have a higher chance of being selected.

I need to rethink the implementations of `fastRand(x)` and `fastRand(x, y)`, since I failed to consider how and where the rounding happens. It probably needs to happen inside these functions.

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-	s->m_startFreqModel.setValue( 0.5f+frnd(0.5f) );
-	s->m_minFreqModel.setValue(	s->m_startFreqModel.value()-0.2f-frnd(0.6f)	);
+	s->m_waveFormModel.setValue(fastRand(2));
+	if (s->m_waveFormModel.value() == 2 && oneIn(2)) { s->m_waveFormModel.setValue(fastRand() & 1); }
```
An "and 1" parity check spotted in the wild! :smile:

rubiefawn replied:
Less of a parity check, and more of a faster `% 2` to randomly select between values of 0 and 1. This technically prefers 1 over 0 a tiny amount (0.003%) of the time lol

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-	s->m_slideModel.setValue( -0.15f-frnd(0.2f) );
-
-	if(rnd(2)==0)
+	if (oneIn(2))
```
IIRC, `if (rnd(2) == 0)` corresponds to `oneIn(3)`.

rubiefawn replied:
Good catch!

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-		s->m_sqrDutyModel.setValue( 0.4f+frnd(0.5f) );
-		s->m_sqrSweepModel.setValue( -frnd(0.7f) );
+		s->m_sqrDutyModel.setValue(fastRand(0.4f, 0.9f));
+		s->m_sqrSweepModel.setValue(fastRand(-0.7f));
```
Since `fastRand(range)` actually multiplies `fastRand()`, `range` and the 1/32767 fraction, this is identical to just having the sign in front, so it's all good!

I also like the encapsulation of the sign flip logic in the actual function argument list.

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-		s->m_startFreqModel.setValue( 0.2f+frnd(0.7f) );
-		s->m_slideModel.setValue( -0.2f-frnd(0.2f) );
+		s->m_startFreqModel.setValue(fastRand(0.2f, 0.9f));
+		s->m_slideModel.setValue(fastRand(-0.2f, -0.4f));
```
The arguments should be flipped, since the first is smaller.

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-
-	s->m_startFreqModel.setValue( 0.2f+frnd(0.6f) );
-	s->m_slideModel.setValue( -0.3f-frnd(0.4f) );
+	s->m_waveFormModel.setValue(fastRand(2));
```
Again, `fastRand() % 3`.

rubiefawn replied:
This should no longer be relevant as of f6ae0f75c22e351f71c1df5e235271b8c830dea0.

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-		s->m_hpFilCutModel.setValue( frnd(0.3f) );
-	}
+	s->m_startFreqModel.setValue(fastRand(0.2f, 0.8f));
+	s->m_slideModel.setValue(fastRand(-0.3f, -0.7f));
```
Again, argument flipping.

bratpeki commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-	s->m_lpFilCutSweepModel.setValue(std::pow(frnd(2.0f) - 1.0f, 3.0f));
-	if(s->m_lpFilCutModel.value()<0.1f && s->m_lpFilCutSweepModel.value()<-0.05f)
+	s->m_lpFilResoModel.setValue(fastRand(-1.f, +1.f));
+	s->m_lpFilCutModel.setValue(1.0f - pow_hack<3>(fastRand(-1.f, +1.f)));
```
This is nauseatingly many LOC, but I'm fairly certain the original is:
```cpp
s->m_lpFilCutModel.setValue(1.0f - std::pow(frnd(1.0f), 3.0f)); // LOC 1039
```

`frnd(1.0f)` was swapped out for `fastRand(-1.f, +1.f)`, instead of `fastRand(1.f)`.

rubiefawn replied:
You are correct, good catch! I appear to have gotten `m_lpFilCutModel` confused with `m_lpFilCutSweepModel`.

rubiefawn committed:
Fix some code review stuff

rubiefawn committed:
Fix faulty random range math

bratpeki commented on plugins/Vibed/VibratingString.cpp:
```diff
@@ -40,7 +40,7 @@ VibratingString::VibratingString(float pitch, float pick, float pickup, const fl
 	m_randomize{randomize},
 	m_stringLoss{1.0f - stringLoss},
-	m_choice{static_cast<int>(m_oversample * static_cast<float>(std::rand()) / static_cast<float>(RAND_MAX))},
+	m_choice(fastRand(m_oversample)),
```
Had to verify the types, so I'll just leave this as reference for future checks.

https://github.com/rubiefawn/lmms/blob/fix/rand/plugins/Vibed/VibratingString.h#L87

`m_oversample` is an int, and the original code does `rand()/RAND_MAX`, which covers the range [0, 1], so the rewrite is correct.

rubiefawn committed:
Remove plus signs per review comments

rubiefawn commented on include/lmms_math.h:
```diff
@@ -176,26 +195,24 @@ inline float linearToLogScale(float min, float max, float value)
 inline auto fastPow10f(T x)
 {
-	return std::exp(std::numbers::ln10_v<T> * x);
+	return std::exp(ln10_v<std::conditional_t<std::is_floating_point_v<T>, T, float>> * x);
```
> Why do this at all? Why not use exclusively [`ln10`](https://en.cppreference.com/w/cpp/numeric/constants.html), which seems to be a `double`, or `ln10_v<float>`? Is it performance-related?

Kind of. I'm not sure how big of a difference it makes as I haven't benchmarked it, but it boils down to eliminating unnecessary casts between `float` and `double`. Take this example code ([godbolt link](https://godbolt.org/z/arosPn999)):

```cpp
#include <cmath>
#include <numbers>
#include <concepts>

template<typename T> requires std::is_arithmetic_v<T>
auto fastPow10f_no_conditional_type(T x) {
	return std::exp(std::numbers::ln10 * x);
}

template<typename T> requires std::is_arithmetic_v<T>
auto fastPow10f(T x) {
	return std::exp(std::numbers::ln10_v<std::conditional_t<std::is_floating_point_v<T>, T, float>> * x);
}

int main() {
	// No casts required; double comes in, double goes out
	double a = fastPow10f(123.45);
	// No casts required; double comes in, double goes out
	double b = fastPow10f_no_conditional_type(123.45);

	// No casts required; float comes in, float goes out
	float c = fastPow10f(123.45f);
	// Two casts required; float goes in, gets promoted to a double,
	// calculations happen, then result is cast to float again
	float d = fastPow10f_no_conditional_type(123.45f);
}
```

Inspecting output assembly does indeed show that using the conditional type eliminates up to two `cvtss2sd` (x86)/`fcvt` (ARM) instructions per function call.

If you're reading the example code and thinking, "Just use `ln10_v<float>` instead of `ln10`!", you'd be mostly correct. The vast majority of the call-sites for this function pass in a `float`. However, using `ln10_v<float>` would mean that all these same casts would be happening in every place where this function is called with a `double` instead.

This seemed like an avoidable problem to me, hence the conditional type. It may be ugly, but it is still comprehensible, and it does allow both `float` and `double` to be used without compromise.

rubiefawn committed:
Fix some lmms::fastPow10f() callsites

Mostly casting issues

rubiefawn commented on plugins/SpectrumAnalyzer/SaSpectrumView.cpp:
```diff
@@ -728,17 +728,18 @@ std::vector<std::pair<float, std::string>> SaSpectrumView::makeLogAmpTics(int lo
+	if ((high - low) < 20 * heightScale)
 	{
-		increment = fastPow10f(0.3f); // 3 dB steps when really zoomed in
+		increment = fastPow10f(0.3); // 3 dB steps when really zoomed in
```
`increment` is a `double`. Omitting the 'f' makes `0.3` a `double` literal. Keeping it results in an implicit cast.

rubiefawn commented on include/lmms_math.h:
```diff
@@ -176,26 +195,24 @@ inline float linearToLogScale(float min, float max, float value)
+inline auto fastLog10f(T x)
 {
-	constexpr auto inv_ln10 = static_cast<float>(1.0 / std::numbers::ln10);
+	constexpr auto inv_ln10 = static_cast<std::conditional_t<std::is_floating_point_v<T>, T, float>>(1.0 / ln10);
```
compile time 👍

bratpeki commented on plugins/Organic/Organic.cpp:
```diff
@@ -231,71 +231,63 @@ void OrganicInstrument::playNote( NotePlayHandle * _n,
-			static_cast<oscPtr *>( _n->m_pluginData )->phaseOffsetRight[i]
-				= rand() / (static_cast<float>(RAND_MAX) + 1.0f);
+			newOsc->phaseOffsetLeft[i] = fastRand(1.f);
+			newOsc->phaseOffsetRight[i] = fastRand(1.f);
```
[`rand()`](https://cplusplus.com/reference/cstdlib/rand/) range: [[0, RAND_MAX]](https://www.cs.yale.edu/homes/aspnes/pinewiki/C(2f)Randomization.html#:~:text=inclusive)

Therefore, since we're at maximum getting `RAND_MAX / RAND_MAX + 1`, isn't the range somewhat less than 1? Maybe `fastRand(1.f - 1.f / RAND_MAX)`?

rubiefawn committed:
Re-allow 100% note skip chance

lol

rubiefawn committed:
Technically correct random phase for Organic oscs

Will anybody ever notice this in practice? Probably not lol

rubiefawn committed:
Re-allow 100% arpeggio miss rate

szeli1 commented on include/lmms_math.h:
```diff
@@ -163,7 +182,7 @@ inline float logToLinearScale(float min, float max, float value)
 inline float linearToLogScale(float min, float max, float value)
 {
-	constexpr auto inv_e = static_cast<float>(1.0 / std::numbers::e);
+	constexpr auto inv_e = static_cast<float>(1.0 / e);
```
Why did you choose to use `using namespace std::numbers;`?

szeli1 commented on plugins/Sid/SidInstrument.cpp:
```diff
@@ -238,54 +238,46 @@ float SidInstrument::desiredReleaseTimeMs() const
+	int badline = fastRand(NUMSIDREGS - 1);
+	int residdelay = 0; // customly added
 
+	for (int c = 0; c < NUMSIDREGS; c++)
```
```suggestion
	for (int c = 0; c < NUMSIDREGS; ++c)
```

szeli1 commented on src/core/AutomatableModel.cpp:
```diff
@@ -429,37 +429,26 @@ void AutomatableModel::setStep( const float step )
+	if (m_step != 0 && m_hasStrictStepSize)
 	{
-		value = nearbyintf( value / m_step ) * m_step;
+		value = std::nearbyintf(value / m_step) * m_step;
```
What if the rounding direction changes?

rubiefawn replied:
Evil things, I imagine. I don't think `nearbyintf()` is wise here either, but addressing that probably belongs in its own PR, so I have opted to preserve the existing behavior here.

Veratil replied:
I agree with @rubiefawn in this case that we should address this in another PR. Nothing is changing here from a behavior point of view, just adding the std namespace.

rubiefawn replied:
The only reason I made any changes to this file was due to some rounding behavior in `fastRand()`, which as since been resolved. I have reverted all changes to this file in this branch as of 79f8bdb2a8020b59f1c610eece5f781365f381f0. Please direct any further discussion about `nearbyintf()` to #8119.

szeli1 reviewed (dismissed):
I like the code. I did not test this.

rubiefawn committed:
Pre-increment because it looks nice

rubiefawn committed:
Remove `using namespace std::numbers`

rubiefawn commented on src/core/InstrumentFunctions.cpp:
```diff
@@ -414,28 +415,20 @@ void InstrumentFunctionArpeggio::processNote( NotePlayHandle * _n )
 
 		// Skip notes randomly
-		if( m_arpSkipModel.value() )
+		if (m_arpSkipModel.value() && fastRand(100.f) <= m_arpSkipModel.value())
```
It looks exclusive, but note the `<` instead of `<=`. The old code was inclusive too, just in a weird way.

Additionally, the knob does in fact go up to 100%.

[See also](https://github.com/LMMS/lmms/pull/7741#discussion_r2449372130)

rubiefawn commented on include/lmms_math.h:
```diff
@@ -81,26 +82,43 @@ inline auto absFraction(std::floating_point auto x) noexcept
 }
 
+
+//! @brief Returns a psuedorandom number between @p from and @p to, inclusive.
```
What is the problem with using swapped values? It should be the same?

bratpeki replied:
`fastRand(range)` excepts negatives?

```cpp
fastRand(-15) =
  floor(fastRand() * (1 + -15) * FAST_RAND_RATIO) =
  floor(fastRand() * (-14) * FAST_RAND_RATIO)
```

The range is between -14 and 0, right?

rubiefawn replied:
God damn it, the rounding comes back to haunt me once again. Fixed (I hope) in e50b6614eb5b8f8704bc078b31611787cef9f413 with some detailed comments.

rubiefawn commented on plugins/Sfxr/Sfxr.cpp:
```diff
@@ -745,359 +735,307 @@ void SfxrInstrumentView::modelChanged()
-	s->m_slideModel.setValue( -0.15f-frnd(0.2f) );
-
-	if(rnd(2)==0)
+	if (oneIn(3))
```
```cpp
// This is how rnd() was defined before I removed it:
#define rnd(n) (rand()%(n+1))

// This is what that if statement would look like with the macro expanded:
if(rnd(2)==0) {...}
if((rand()%(3))==0) {...}

// This is exactly how oneIn() is currently defined:
inline bool oneIn(unsigned chance) noexcept { return 0 == (fastRand() % chance); }

// So in conclusion, this:
if((rand()%(3))==0) {...}
// Becomes this:
if (oneIn(3)) {...}
```

bratpeki replied:
Yup, it's just a misalignment! Use the else block and the like as reference to see what pairs up with what.

rubiefawn committed:
Fix more evil rounding

Also Github's diff viewer is cursed and I am now filled with distrust
towards it

messmerd commented on include/lmms_math.h:
```diff
@@ -180,26 +208,24 @@ inline float linearToLogScale(float min, float max, float value)
 inline auto fastPow10f(T x)
 {
-	return std::exp(std::numbers::ln10_v<T> * x);
+	return std::exp(std::numbers::ln10_v<std::conditional_t<std::is_floating_point_v<T>, T, float>> * x);
```
Or just have two functions:
```cpp
template<std::floating_point T>
inline T fastPow10f(T x)
{
	return std::exp(std::numbers::ln10_v<T> * x);
}

template<std::integral T>
inline float fastPow10f(T x)
{
	return fastPow10f(static_cast<float>(x));
}
```

rubiefawn committed:
Merge branch 'master' into fix/rand

rubiefawn committed:
Extract conditional type for readability

szeli1 reviewed (commented):
I'm changing my approval in light of `std::nearbyint`.

rubiefawn committed:
Revert changes to src/core/AutomatableModel.cpp

Out of scope, these belong on a different branch

rubiefawn committed:
Remove old commented-out debug statement

regulus79 reviewed (approved):
I was asked to review this PR; I have looked through all the code, and the changes appear fine to me. I am not an expert in pseudorandom number generation, but assuming people have tested it, I think it's fine.

messmerd commented on include/lmms_math.h:
```diff
@@ -81,26 +82,53 @@ inline auto absFraction(std::floating_point auto x) noexcept
+	const float frange = static_cast<float>(range); // Even on -O3 it casts twice without this for some reason??
+	const float r = frange + std::copysign(1.f, frange);
+	// Always round towards 0 (implicit truncation occurs during static_cast).
+	return static_cast<T>(fastRand() * r * FAST_RAND_RATIO);
```
```suggestion
//! @brief Returns a pseudorandom integer in the range [0, @p upper)
template<std::integral T>
inline T fastRand(T upper) noexcept
{
	// The integer specialization of this function is kind of weird, but
	// it is necessary to prevent massive bias away from the maximum
	// value. FAST_RAND_RATIO here is 1 greater than normal, so it will
	// actually result in an open-ended range.
	constexpr float FAST_RAND_RATIO = 1.f / 32768;
	// Since it's open-ended using the above ratio, increase the magnitude
	// by 1. All values greater than upper get rounded to upper, and
	// the bias is removed.
	const float fupper = static_cast<float>(upper); // Even on -O3 it casts twice without this for some reason??
	const float r = fupper + std::copysign(1.f, fupper);
	// Always round towards 0 (implicit truncation occurs during static_cast).
	return static_cast<T>(fastRand() * r * FAST_RAND_RATIO);
```
Some improvements + typo fixes

messmerd commented on include/lmms_math.h:
```diff
@@ -81,26 +82,53 @@ inline auto absFraction(std::floating_point auto x) noexcept
 }
 
+
+//! @brief Returns a psuedorandom number between @p from and @p to, inclusive.
```
```suggestion
//! @brief Returns a pseudorandom number in the range [@p from, @p to] (inclusive)
```

messmerd commented on plugins/Organic/Organic.cpp:
```diff
@@ -231,71 +231,66 @@ void OrganicInstrument::playNote( NotePlayHandle * _n,
+			// TODO C++23: Use std::nextafter(1.f, 0.f) since it will be constexpr
+			constexpr float MAX_RANDOM_PHASE = 0.999999940395355224609375f;
+			newOsc->phaseOffsetLeft[i] = fastRand(MAX_RANDOM_PHASE);
+			newOsc->phaseOffsetRight[i] = fastRand(MAX_RANDOM_PHASE);
```
I don't really like this. Maybe we could have a `fastRandEx` function (or similar name) whose upper bound is exclusive?

rubiefawn replied:
Implemented as of 0e4ee0e51079404a4151931e9c5093137b1cec12

messmerd commented on plugins/Sid/SidInstrument.cpp:
```diff
@@ -238,54 +238,46 @@ float SidInstrument::desiredReleaseTimeMs() const
-  return total;
-}
+	int total = 0;
+	int badline = fastRand(NUMSIDREGS - 1);
```
This could use `fastRandEx` if it's added.

messmerd reviewed (commented):
Quick review - haven't looked at things too closely yet

rubiefawn committed:
Pass reference to temporary copy

oopsie daisy

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

rubiefawn committed:
Clarify comments and documentation

Also use "upper" instead of "range" for fastRand(x) functions, since the
lone parameter is not itself a range, but rather the upper limit of one.

Today I learned that I do not know how to spell "pseudorandom" lmao

rubiefawn committed:
Simplify template parameter constraints

Didn't know you could just put em in the angle brackets lol, neat

rubiefawn committed:
Inclusive and exclusive upper bound variants

By popular demand!
